spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/RemoveNetCarryoverIterator.java

```diff
@@ -0,0 +1,143 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.util.Iterator;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * This class computes the net changes across multiple snapshots. It takes a row iterator, and
+ * assumes the following:
+ *
+ * <ul>
+ *   <li>The row iterator is partitioned by all columns.
+ *   <li>The row iterator is sorted by all columns, change order, and change type. The change order
+ *       is 1-to-1 mapping to snapshot id.
+ * </ul>
+ */
+public class RemoveNetCarryoverIterator extends RemoveCarryoverIterator {
+
+  private final int[] indicesToIdentifySameRow;
+
+  private Row cachedNextRow = null;
+  private Row cachedRow = null;
+  private long cachedRowCount = 0;
+
+  protected RemoveNetCarryoverIterator(Iterator<Row> rowIterator, StructType rowType) {
+    super(rowIterator, rowType);
+    this.indicesToIdentifySameRow = generateIndicesToIdentifySameRow();
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRowCount > 0) {
+      return true;
+    }
+
+    if (cachedNextRow != null) {
+      return true;
+    }
+
+    return rowIterator().hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there are cached rows, return one of them from the beginning
+    if (cachedRowCount > 0) {
+      cachedRowCount--;
+      return cachedRow;
+    }
+
+    Row currentRow = getCurrentRow();
+
+    // if there is no next row, return the current row
+    if (!rowIterator().hasNext()) {
+      return currentRow;
+    }
+
+    Row nextRow = rowIterator().next();
+
+    cachedRow = currentRow;
+    cachedRowCount = 1;
+    // if they are the same row, remove the cached row or stack to the cached row
+    while (isSameRecord(currentRow, nextRow)) {
+      if (matched(currentRow, nextRow)) {
+        // remove both rows
+        cachedRowCount--;
+        nextRow = null;
+      } else {
+        // stack the next row to the cached row
+        nextRow = null;
+        cachedRowCount++;
+      }
+
+      // break the loop if there is no next row or the cache is empty
+      if (cachedRowCount <= 0 || !rowIterator().hasNext()) {
+        break;
+      }
+
+      nextRow = rowIterator().next();
+    }
+
+    // if they are different rows, hit the boundary, cache the next row
+    cachedNextRow = nextRow;
+    return null;
+  }
+
+  private Row getCurrentRow() {
+    Row currentRow;
+    if (cachedNextRow != null) {
+      currentRow = cachedNextRow;
+      cachedNextRow = null;
+    } else {
+      currentRow = rowIterator().next();
+    }
+    return currentRow;
+  }
+
+  private boolean matched(Row currentRow, Row nextRow) {
```

Comment:
Neds a better name since this isn't "matched" this is whether or not the pair of rows represent a net change correct?